# **EDA on NETFLIX 🍿🎥**

<img src="./images/Netflix_Homepage.jpg" alt="Image Description">

---
**Author:** `Syed Ghazi Ali Zaidi`

* Contact: _sghazializaidi@gmail.com_
* Explore my code: _https://github.com/Ghazi-work_
* Connect with me: _https://www.linkedin.com/in/syed-ghazi-ali-zaidi-405931217_


---

 ## **Data Overview 📊**

* **Attributes:** Explore key details such as show ID, type (movie or TV show), title, director, cast, country, date added, release year, rating, duration, and more.
* **Record Structure:** Each row unfolds a unique show, providing a snapshot of its cinematic attributes.

| Column        | Data Type | Description |
|---------------|-----------|-------------|
| show_id       | Object    | Identifier for the show/movie |
| type          | Object    | Type of the content (Movie or TV Show) |
| title         | Object    | Title of the show |
| director      | Object    | Director of the show/movie (if available) |
| cast          | Object    | Cast members of the show/movie |
| country       | Object    | Country where the show/movie was produced |
| date_added    | Object    | Date when the show/movie was added to Netflix |
| release_year  | Numeric   | Year when the show/movie was released |
| rating        | Object    | Content rating of the show/movie |
| duration      | Object    | Duration of the show/movie |
| listed_in     | Object    | Categories in which the show/movie is listed |
| description   | Object    | Brief description of the show/movie |

## **Motivation 🚀** 

While movie databases are aplenty, exploring the Netflix galaxy provides a unique perspective. This dataset opens doors to analyze content trends, viewer preferences, and the global impact of Netflix's vast library. Gathering this celestial data had its challenges, making it an even more intriguing resource.


## **Credits and Acknowledgements 🌟**

A heartfelt gratitude to the data curators at Netflix and the original contributors on [Kaggle](https://www.kaggle.com/shivamb/netflix-shows). Their dedication has bestowed upon us this remarkable dataset, opening doors to endless possibilities.

Special thanks to [Dr. Aammar Tufail](https://github.com/AammarTufail), whose guidance has been a beacon in the sea of data exploration.

Let the analysis begin! 🚀📊

#### **Kernel Version**:
* Python 3.11.5

# **1. Importing Libraries**

In [42]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import plotly
# connected=True means it will download the latest version of plotly javascript library.
plotly.offline.init_notebook_mode(connected=True)
import plotly.graph_objs as go
import plotly.figure_factory as ff
import plotly.express as px
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')
pd.option_context('display.max_columns', None)

# **2. Loading Dataset and Exploring**

In [2]:
df = pd.read_csv('./Datasets/Netflix_data.csv')

* Sneak peek at the dataset

In [3]:
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [4]:
print(f'There are {df.shape[0]} rows and {df.shape[1]} columns in the dataset.')

There are 8807 rows and 12 columns in the dataset.


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


> ##### 1. The column `show_id` can be numeric after removing the character `s`.
> ##### 2. The column `date_added` can be divided in to different column like `day`,`month`, and `year` for better analysis.



#### Looking at selected columns to find something intresting 🔎

In [6]:
df['duration'].value_counts()

duration
1 Season     1793
2 Seasons     425
3 Seasons     199
90 min        152
94 min        146
             ... 
16 min          1
186 min         1
193 min         1
189 min         1
191 min         1
Name: count, Length: 220, dtype: int64

* **Intresting** the `duration` column can be used to make **2** different dataframes one for only **movies** that holds time and seperate for **Seasons** that holds season count. Help us get good insights 🤩

In [7]:
df['rating'].value_counts()

rating
TV-MA       3207
TV-14       2160
TV-PG        863
R            799
PG-13        490
TV-Y7        334
TV-Y         307
PG           287
TV-G         220
NR            80
G             41
TV-Y7-FV       6
NC-17          3
UR             3
74 min         1
84 min         1
66 min         1
Name: count, dtype: int64

* **Hmmmm 🤔** looks like some other column data has replaced the `rating` data which needs to be handled in the next phase.

* Also the `rating` column data doesn't communicate much.. I'll try to give meaningful names to each.\
like: PG as *Parents Guidance*

In [8]:
df['country'].value_counts()

country
United States                             2818
India                                      972
United Kingdom                             419
Japan                                      245
South Korea                                199
                                          ... 
Romania, Bulgaria, Hungary                   1
Uruguay, Guatemala                           1
France, Senegal, Belgium                     1
Mexico, United States, Spain, Colombia       1
United Arab Emirates, Jordan                 1
Name: count, Length: 748, dtype: int64

* Looks like some movie/shows have **more than 1** country listed, it can be due to collaborative work between two or more countries

In [9]:
df['listed_in'].value_counts()

listed_in
Dramas, International Movies                          362
Documentaries                                         359
Stand-Up Comedy                                       334
Comedies, Dramas, International Movies                274
Dramas, Independent Movies, International Movies      252
                                                     ... 
Kids' TV, TV Action & Adventure, TV Dramas              1
TV Comedies, TV Dramas, TV Horror                       1
Children & Family Movies, Comedies, LGBTQ Movies        1
Kids' TV, Spanish-Language TV Shows, Teen TV Shows      1
Cult Movies, Dramas, Thrillers                          1
Name: count, Length: 514, dtype: int64

### **Observations:** 
----
1. There are 8807 rows and 12 columns.

2. The columns in the dataset are:
    - **`'show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'listed_in', 'description'`**

#### **Operations to be Performed ⚒️**
 1. `show_id` which can be numeric if we remove `s` 
 
 2. `date_added` can be divided into 3 columns. `Year`,`month`, and `Day`
 3. `rating` data has to be labeled into more meaningful way, also handle the replaced values by duration.
 4.  `duration` **2** dataframes will be made for each:
        - Movie: This dataframe will hold `duration` in *Hours*
        - Show: This dataframe will hold `duration` as *Season count*
----

# **3. Data Pre-processing**

* In this section we'll do cleaning, transforming, and integrating of the Netflix dataframe, the above mention operations will be performed step by step. So sit back and watch... 😃

#### **1. Show_id**

* **Recap:** character `s` has to be removed later convert it into int64.

* **Reason:** After converting to *Numeric* aggeragate functions can be applied on this column.

* checking null values

In [10]:
df['show_id'].isnull().sum()

0

In [11]:
print(df['show_id'].unique())

print('---------------------') # seperator

df['show_id'].value_counts()

['s1' 's2' 's3' ... 's8805' 's8806' 's8807']
---------------------


show_id
s1       1
s5875    1
s5869    1
s5870    1
s5871    1
        ..
s2931    1
s2930    1
s2929    1
s2928    1
s8807    1
Name: count, Length: 8807, dtype: int64

> ##### * It shows that the values are indeed `unique id` of the show/movie.

* Replacing `s` with empty space

In [12]:
df['show_id'] = df['show_id'].str.replace('s','') # replacing 's'

df['show_id'] = df['show_id'].astype('int64') # Converting it to int64 

df['show_id'].info() # checking the data type after converting

<class 'pandas.core.series.Series'>
RangeIndex: 8807 entries, 0 to 8806
Series name: show_id
Non-Null Count  Dtype
--------------  -----
8807 non-null   int64
dtypes: int64(1)
memory usage: 68.9 KB


#### **2. Date_added**

* **Recap:** The date will divided into 3 columns `Year`, `Month`, and `Day`

* **Reason:** Would be helpful in generating insights even on `Day` and `Month` level

- Viewing to understanding the pattern

In [13]:
df['date_added'].sample(5)

5884    November 13, 2015
7826        June 11, 2019
8395     November 6, 2020
535          July 6, 2021
2710        April 3, 2020
Name: date_added, dtype: object

> * Month first, then the day in last we get the year

In [14]:
# Converting the datatype of 'date_added' to date time
df['date_added'] = pd.to_datetime(df['date_added'], errors='coerce')

# Extracting year, month, and day into separate columns
df['year_added'] = df['date_added'].dt.year
df['month_added'] = df['date_added'].dt.month
df['day_added'] = df['date_added'].dt.day

* Viewing the first 5 rows after operation

In [15]:
df[['date_added','year_added','month_added','day_added']].head()

date_added  year_added  month_added  day_added
0 2021-09-25      2021.0          9.0       25.0
1 2021-09-24      2021.0          9.0       24.0
2 2021-09-24      2021.0          9.0       24.0
3 2021-09-24      2021.0          9.0       24.0
4 2021-09-24      2021.0          9.0       24.0

#### **3. Rating**

* **Recap:** This column data has to be converted into more meaningful way, secondly there were issues with 3 rows.

* **Reason:** Would be helpful in understanding when presenting the data

In [16]:
df['rating'].isnull().sum()

4

> * We will be dropping the nulls as they are few in numbers and cannot be imputed

In [17]:
df['rating'].value_counts()

rating
TV-MA       3207
TV-14       2160
TV-PG        863
R            799
PG-13        490
TV-Y7        334
TV-Y         307
PG           287
TV-G         220
NR            80
G             41
TV-Y7-FV       6
NC-17          3
UR             3
74 min         1
84 min         1
66 min         1
Name: count, dtype: int64

#### We'll Print those rows having `74 min`, `84 min` and `66 min` in `rating` column

In [18]:
# Identify rows with the specified previous 'rating' values
wrong_ratings = ['74 min', '84 min', '66 min']
modified = df[df['rating'].isin(wrong_ratings)]

# Print the relevant columns for these rows
modified[['show_id', 'rating', 'duration']].head()


show_id  rating duration
5541     5542  74 min      NaN
5794     5795  84 min      NaN
5813     5814  66 min      NaN

> * As can be seen that duration values have come in place of rating values, even if we swap the values we still won't be able to judge the `rating` so the best way is to remove these rows

In [19]:
show_ids_to_remove = [5542, 5795, 5814]
df = df.dropna(subset=['rating']) # dropping nan from rating
df = df[~df['show_id'].isin(show_ids_to_remove)] # removing rows with specified ids

* Now giving meaningful name to rating values. 🙂

In [20]:
df['rating'].unique() # Viewing unique values in the rating

array(['PG-13', 'TV-MA', 'PG', 'TV-14', 'TV-PG', 'TV-Y', 'TV-Y7', 'R',
       'TV-G', 'G', 'NC-17', 'NR', 'TV-Y7-FV', 'UR'], dtype=object)

* Mapping each column with meaningful names

In [21]:
# Mapping of original ratings to meaningful names
rating_mapping = {
    'PG-13': 'Parental Guidance 13+',
    'TV-MA': 'Mature Audiences',
    'PG': 'Parental Guidance',
    'TV-14': 'Parents Cautioned 14+',
    'TV-PG': 'Parental Guidance',
    'TV-Y': 'All Children',
    'TV-Y7': 'Directed to Older Children',
    'R': 'Restricted',
    'TV-G': 'General Audiences',
    'G': 'General Audiences',
    'NC-17': 'No one 17 and under admitted',
    'NR': 'Not Rated',
    'TV-Y7-FV': 'Fantasy Violence',
    'UR': 'Unrated'
}

# Create a new column 'rating_meaningful' based on the mapping
df['rating_meaningful'] = df['rating'].map(rating_mapping)

# Print the DataFrame
df[['rating', 'rating_meaningful']].sample(6).head()

rating      rating_meaningful
8395     PG      Parental Guidance
1647  TV-14  Parents Cautioned 14+
3963  TV-MA       Mature Audiences
2963   TV-G      General Audiences
313   TV-14  Parents Cautioned 14+

### **Handling Duplicates**

In [22]:
df.duplicated().sum()

0

> * There seems no duplicated values but let me check `description` column

In [23]:
df['description'].duplicated().sum()

32

> * 🫣 32 duplicates.. There seems to have dublicated values in the dataset, let's check which rows are they..

In [24]:
df[['show_id','title','description','date_added','release_year']].loc[df['description'].duplicated()].head().sort_index()

show_id                    title  \
79        80  Tughlaq Durbar (Telugu)   
237      238          Boomika (Hindi)   
238      239      Boomika (Malayalam)   
239      240         Boomika (Telugu)   
851      852         99 Songs (Tamil)   

                                           description date_added  \
79   A budding politician has devious plans to rise... 2021-09-11   
237  Paranormal activity at a lush, abandoned prope... 2021-08-23   
238  Paranormal activity at a lush, abandoned prope... 2021-08-23   
239  Paranormal activity at a lush, abandoned prope... 2021-08-23   
851  Challenged to compose 100 songs before he can ... 2021-05-21   

     release_year  
79           2021  
237          2021  
238          2021  
239          2021  
851          2021

> * As can be seen that there are movies that are same but in different languge, we'll be taking one of their instances.

##### Droping duplicated rows through `description` column 😀

In [25]:
df = df.drop_duplicates(subset='description') 
df['description'].duplicated().sum() # checking if duplicates were removed

0

### **Handling Nulls**

In [26]:
df.isnull().sum().sort_values(ascending=False)   

director             2623
cast                  820
country               818
date_added             97
year_added             97
month_added            97
day_added              97
show_id                 0
type                    0
title                   0
release_year            0
rating                  0
duration                0
listed_in               0
description             0
rating_meaningful       0
dtype: int64

In [27]:
df = df.dropna(subset=['date_added','year_added','month_added','day_added','director','cast','country'])

### **Visualization of the data📊**

-----

### *1. Content type on Netflix*

In [44]:
type_count = df['type'].value_counts()

# Calculate percentages
percentages = (type_count / type_count.sum()) * 100

# pie chart using plotly
fig = px.pie(names=percentages.index, values=percentages.values, 
             title='Content Type on Netflix', 
             labels=percentages.index, color_discrete_sequence=['skyblue', 'lightcoral'],
             hole=0.3, opacity=0.7)
fig.update_traces(textinfo='percent+label', pull=[0, 0.1])  # Adjust labels and pull a slice

fig.show()


### *2. Trend of content over the years*

In [48]:
type_counts = df.groupby(['year_added', 'type']).size().reset_index(name='count')

# line chart using plotly
fig = px.line(type_counts, x='year_added', y='count', color='type',
              labels={'count': 'Count', 'year_added': 'Year Added'},
              title='Trend of Movies and TV Shows Over the Years')

fig.show()

In [54]:
df.head()

show_id     type                          title             director  \
7         8    Movie                        Sankofa         Haile Gerima   
8         9  TV Show  The Great British Baking Show      Andy Devonshire   
9        10    Movie                   The Starling       Theodore Melfi   
12       13    Movie                   Je Suis Karl  Christian Schwochow   
24       25    Movie                          Jeans           S. Shankar   

                                                 cast  \
7   Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
8   Mel Giedroyc, Sue Perkins, Mary Berry, Paul Ho...   
9   Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12  Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
24  Prashanth, Aishwarya Rai Bachchan, Sri Lakshmi...   

                                              country date_added  \
7   United States, Ghana, Burkina Faso, United Kin... 2021-09-24   
8                                      United Kingdom 2021-09-24   
9                                       United States 2021-09-24   
12                            Germany, Czech Republic 2021-09-23   
24                                              India 2021-09-21   

    release_year rating   duration  \
7           1993  TV-MA    125 min   
8           2021  TV-14  9 Seasons   
9           2021  PG-13    104 min   
12          2021  TV-MA    127 min   
24          1998  TV-14    166 min   

                                           listed_in  \
7   Dramas, Independent Movies, International Movies   
8                       British TV Shows, Reality TV   
9                                   Comedies, Dramas   
12                      Dramas, International Movies   
24   Comedies, International Movies, Romantic Movies   

                                          description  year_added  \
7   On a photo shoot in Ghana, an American model s...      2021.0   
8   A talented batch of amateur bakers face off in...      2021.0   
9   A woman adjusting to life after a loss contend...      2021.0   
12  After most of her family is murdered in a terr...      2021.0   
24  When the father of the man she loves insists t...      2021.0   

    month_added  day_added      rating_meaningful  
7           9.0       24.0       Mature Audiences  
8           9.0       24.0  Parents Cautioned 14+  
9           9.0       24.0  Parental Guidance 13+  
12          9.0       23.0       Mature Audiences  
24          9.0       21.0  Parents Cautioned 14+

In [61]:
import plotly.express as px

# Assuming your DataFrame is named 'df'
# Convert 'date_added' to datetime if it's not already
df['date_added'] = pd.to_datetime(df['date_added'], errors='coerce')

# Calculate the time gap between release year and date added
df['time_gap'] = df['date_added'].dt.year - df['release_year']

# Create a box plot
fig = px.box(df, y='time_gap',
             labels={'time_gap': 'Time Gap (Years)'},
             title='Box Plot of Time Gaps Between Release Year and Date Added')

# Show the plot
fig.show()
